glusterd2/commands/volumes/volume-profile.go

```diff
@@ -0,0 +1,309 @@
+		} else {
+			// Create new map with fop as key if this particular Fop is encountered for the first time
+			stats.StatsInfo[fop] = make(map[string]string)
+			stats.StatsInfo[fop][k] = value
+		}
+	} else {
+		if k == "read" {
+			stats.DataRead = value
+		} else if k == "write" {
+			stats.DataWrite = value
+		} else if k == "duration" {
+			stats.Duration = value
+		}
+	}
+	return stats
+}
+
+// Decode key for fop and stat. keys starting with -1 belong to cumulative stat
+// Eg: -1-12-maxlatency fop : 12        stat: maxlatency
+// -1-duration          fop : ""        stat: duration
+func decodeCumulativeKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 4 {
+		k = s[len(s)-1]
+		index, _ := strconv.Atoi(s[2])
+		fop = fops[index]
+	} else if len(s) == 3 {
+		k = s[len(s)-1]
+		fop = ""
+	}
+	return fop, k
+}
+
+// Decode key for fop and stat for keys belonging to interval stat
+// Eg: 12-12-maxlatency  IntervalNo.: 12  fop : 12        stat: maxlatency
+// 12-duration           IntervalNo.: 12  fop : ""        stat: duration
+func decodeIntervalKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 3 {
+		k = s[len(s)-1]
+		index, _ := strconv.Atoi(s[1])
+		fop = fops[index]
+	} else if len(s) == 2 {
+		k = s[len(s)-1]
+		fop = ""
+	}
+	return fop, k
+}
+
+func txnVolumeProfile(c transaction.TxnCtx) error {
+	var volinfo volume.Volinfo
+	var nodeProfileInfo []map[string]string
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		return err
+	}
+
+	var option string
+	if err := c.Get("option", &option); err != nil {
+		return err
+	}
+
+	for _, b := range volinfo.GetLocalBricks() {
+		brickDaemon, err := brick.NewGlusterfsd(b)
+		if err != nil {
+			return err
+		}
+
+		c.Logger().WithField(
+			"volume", volinfo.Name).Info("Starting volume profile operation")
+
+		client, err := daemon.GetRPCClient(brickDaemon)
+		if err != nil {
+			c.Logger().WithError(err).WithField(
+				"brick", b.String()).Error("failed to connect to brick, sending SIGTERM")
+			return err
+		}
+		reqDict := make(map[string]string)
+		switch option {
+		case "info":
+			reqDict["peek"] = "0"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "1"
+			reqDict["originator_uuid"] = gdctx.MyUUID.String()
+			break
+		case "info-peek":
+			reqDict["peek"] = "1"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "1"
+			break
+		case "info-incremental":
+			reqDict["peek"] = "0"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "2"
+			break
+		case "info-incremental-peek":
+			reqDict["peek"] = "1"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "2"
+			break
+		case "info-cumulative":
+			reqDict["peek"] = "0"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "3"
+			break
+		case "info-clear":
+			reqDict["peek"] = "0"
+			reqDict["op"] = "3"
+			reqDict["info-op"] = "4"
+			break
+		default:
+			return errors.New("operation not valid")
```

Comment:
an error message can be more specific, return an error message with more info saying which operation is not valid.

Reply:
Done.